Offboard auth

peterbarker opened the pull request.

WickedShell commented on libraries/GCS_MAVLink/#GCS.cpp#:
```diff
@@ -0,0 +1,55 @@
+#include "GCS.h"
```
What is this file? (the name is really weird)

peterbarker replied:
emacs file indicating a file is currently being edited.  Fixed.

WickedShell commented on libraries/GCS_MAVLink/GCS.h.orig:
```diff
@@ -0,0 +1,580 @@
+/// @file	GCS.h
```
Git/patch problems?

peterbarker replied:
Yes.  Fixed.

peterbarker commented:
This seems to be implemented as the "standard" implied by `common.xml` indicates.

However, it looks like the standard may have changed during the implementation of this in PX4 Pro.

The enumeration added isn't used.

There's no way to authorize a specific vehicle to arm with a normal command-ack.

So I'm confused.

magicrub commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -70,6 +70,22 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
+
+    // @Param: OBDUTHTMX
+    // @DisplayName: Offboard Authorization Time Max
+    // @Description: Time authorisation is assumed valid for; 0 to disable
```
spelling authorisation: -s +z

magicrub replied:
"OBDUTHTMX" -> You're missing an A

magicrub commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -70,6 +70,22 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
     AP_GROUPINFO("VOLT2_MIN",     5,     AP_Arming,  _min_voltage[1],  0),
 
+    // @Param: OBAUTHINT
+    // @DisplayName: Offboard Authorization Interval
```
if the param is OB* then lets' have the displayname and desc emphasize the other B like "Off-Board.."

magicrub commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -70,6 +70,22 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
+    // @User: Advanced
+    // @Units: s
+    // @Increment: 0.1
+    AP_GROUPINFO("OBAUTHINT",     6,     AP_Arming,  offboard_auth_interval,  10),
```
Increment: 0.1 is wrong, Should be 1

magicrub commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -658,3 +675,35 @@ bool AP_Arming::rc_checks_copter_sub(const bool display_failure, const RC_Channe
+// expected to be called at some rate to do offboard auth as required
+bool AP_Arming::offboard_authorization_ok(bool report)
+{
+    if (offboard_auth_maxt == 0) {
```
<= 0

magicrub commented:
closes https://github.com/ArduPilot/ardupilot/issues/6405

peterbarker commented:
@magicrub I've addressed the points you raised.

With the exception of the naming of the variable - the A is missing because we're right on the 16 character limit.  The `ARMING_` prefix doesn't help :-)

magicrub commented:
I'd say rename "offboard" to "external" and remove the OB in front of the strings. ARMING_AUTH_*

WickedShell commented on libraries/GCS_MAVLink/GCS_Common.cpp.orig:
```diff
@@ -0,0 +1,2567 @@
+/*
```
more emacs files I guess?

peterbarker commented:
On Thu, 30 Nov 2017, WickedShell wrote:

> more emacs files I guess?

*sigh*.  No, more patch files.  Fixed.

amilcarlucas commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -191,7 +191,10 @@ bool AP_Param::check_group_info(const struct AP_Param::GroupInfo *  group_info,
             return false;
         }
         if (prefix_length + strlen(group_info[i].name) > 16) {
-            Debug("suffix is too long in %s", group_info[i].name);
```
Isn't there a #define somewhere for the 16 ?

auturgy commented:
Can this be a priority please? If it is functional and safe to merge, LGTM.
We know the next step is an alternate method anyway, so debate over naming and the like should perhaps be deferred.

WickedShell commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -2059,7 +2086,9 @@ MAV_RESULT GCS_MAVLINK::handle_command_request_autopilot_capabilities(const mavl
 
 MAV_RESULT GCS_MAVLINK::handle_command_do_send_banner(const mavlink_command_long_t &packet)
 {
+    ::fprintf(stderr, "Sending banner\n");
```
Debug?

peterbarker commented:
This was discussed on the dev call today.

It was discovered that the `COMMAND_ACK` extensions currently present in the mavlink master repository are fundamentally flawed given mavlink2's 0-filling default for extensions (0 is a valid value for target system id / component id).

Since those extensions are required for implementation of this authorization message according to "standard", we simply can not implement this feature using the COMMAND_LONG message.

It was decided that a pair of mavlink2 messages would be created in order to accomplish the same goal.  one message for requesting authorization and a second message granting authorization.

@auturgy We can't merge this after correcting for me not understanding the ack protocol - if we did we'd be locked into mavlink2 extensions on `COMMAND_ACK` which are no good.
